CircleSpawn/CircleSpawn/Views/CircleView.swift

```diff
@@ -0,0 +1,115 @@
+//
+//  CircleView.swift
+//  CircleSpawn
+//
+//  Created by Krystian Duma on 01/04/2020.
+//  Copyright © 2020 DaftAcademy. All rights reserved.
+//
+
+import Foundation
+import UIKit
+
+protocol CircleViewDelegate: class {
+    func bringToFront(subview: CircleView)
+}
+
+class CircleView: UIView {
+    
+    var pan = UIPanGestureRecognizer()
+    var longPress = UILongPressGestureRecognizer()
+    
+    weak var delegate: CircleViewDelegate?
+    
+    init(delegate: CircleViewDelegate, backgroundColor: UIColor, center: CGPoint) {
+        
+        self.delegate = delegate
+        let size: CGFloat = 100
+        super.init(frame: CGRect(origin: .zero, size: CGSize(width: size, height: size)))
+        self.center = center
+        self.backgroundColor = backgroundColor
+        self.layer.cornerRadius = size * 0.5
+        
+        pan.addTarget(self, action: #selector(handlePan(_:)))
+        pan.delegate = self
+        self.addGestureRecognizer(pan)
+        
+        longPress.addTarget(self, action: #selector(handleLongPress(_:)))
+        longPress.delegate = self
+        self.addGestureRecognizer(longPress)
+    }
+
+    func easeIn() {
+        self.alpha = 0
+        self.transform = CGAffineTransform(scaleX: 1.5, y: 1.5)
+        UIView.animate(withDuration: 0.2, animations: {
+            self.alpha = 1
+            self.transform = CGAffineTransform(scaleX: 0.8, y: 0.8)
+        }, completion: {  [weak self] completed in
+            UIView.animate(withDuration: 0.1, animations: {
+                self?.transform = .identity
+            })
+        })
+    }
+
+    func easeOut(completion: ((Bool) -> Void)? = nil) {
+        UIView.animate(withDuration: 0.2, animations: {
+            self.alpha = 0
+            self.transform = CGAffineTransform(scaleX: 2, y: 2)
+        }, completion: completion)
+    }
+    
+    func onSelect() {
+        UIView.animate(withDuration: 0.2, animations: {
+            self.alpha = 0.5
+            self.transform = CGAffineTransform(scaleX: 1.1, y: 1.1)
+        })
+    }
+    
+    func onSelectDismissed() {
+        UIView.animate(withDuration: 0.2, animations: {
+            self.alpha = 1
+            self.transform = CGAffineTransform(scaleX: 1.0, y: 1.0)
+        })
+    }
+    
+    @objc func handlePan(_ pan: UIPanGestureRecognizer) {
+        guard  let view = pan.view else { return }
+        let circle = view as! CircleView
+        let translation = pan.translation(in: view)
+        if longPress.state == .changed {
+            switch pan.state {
+            case .changed:
+                circle.center.x += translation.x
+                circle.center.y += translation.y
+                pan.setTranslation(.zero, in: circle)
+            default:
+                return
+            }
+        }
+    }
```

Comment:
Wydaje mi się że nie powinniśmy przesuwać o tyle ile na swipe robimy (bo jak widać to nie sprawdza się) tylko powinniśmy sprawdzić (o ile to możliwe - strzelam w tej chwili) gdzie znajduje się palec i w to miejsce przesunąć kulkę mając w pamięci relatywny punkt za który złapaliśmy daną kulkę 😇